src/levels.rs

```diff
@@ -1,5 +1,1053 @@
+
+        Ok(new_tables)
+    }
+
+    /// Can allow us to run multiple sub-compactions in parallel across the split key ranges.
+    fn add_splits(&self, compact_def: &mut CompactDef) {
+        compact_def.splits.clear();
+
+        // Pick one every 3 tables.
+        const N: usize = 3;
+
+        // Assume this_range is never inf.
+        let mut skr = compact_def.this_range.clone();
+        skr = skr.extend(&compact_def.next_range);
+
+        let mut add_range = |splits: &mut Vec<KeyRange>, right| {
+            match &mut skr {
+                KeyRange::Range {
+                    left: _,
+                    right: skr_right,
+                } => {
+                    *skr_right = right;
+                }
+                KeyRange::Inf => {
+                    unreachable!()
+                }
+                KeyRange::Empty => {
+                    skr = KeyRange::Range {
+                        left: Bytes::new(),
+                        right,
+                    }
+                }
+            }
+
+            splits.push(skr.clone());
+
+            match &mut skr {
+                KeyRange::Range { left, right } => {
+                    *left = right.clone();
+                }
+                _ => unreachable!(),
+            }
+        };
+
+        for (i, table) in compact_def.bot.iter().enumerate() {
+            if i == compact_def.bot.len() - 1 {
+                add_range(&mut compact_def.splits, Bytes::new());
+                return;
+            }
+
+            if i % N == N - 1 {
+                let biggest = table.biggest();
+                // TODO: Check this.
+                let mut buf = BytesMut::with_capacity(biggest.len() + 8);
+                buf.put(user_key(biggest));
+                let right = key_with_ts(buf, std::u64::MAX);
+                add_range(&mut compact_def.splits, right);
+            }
+        }
+    }
+
+    fn fill_tables_l0_to_l0(&self, compact_def: &mut CompactDef) -> Result<()> {
+        if compact_def.compactor_id != 0 {
+            return Err(Error::CustomError(
+                "only compactor zero can compact L0 to L0".to_string(),
+            ));
+        }
+
+        // TODO: should compact_def be mutable?
+        compact_def.next_level = self.levels[0].clone();
+        compact_def.next_level_id = 0;
+        compact_def.next_range = KeyRange::default();
+        compact_def.bot = vec![];
+
+        let this_level = compact_def.this_level.read().unwrap();
+        // Since next_level and this_level are same L0, do not need to acquire lock of next level.
+        // TODO: Don't lock cpt_status through this function.
+        let mut cpt_status = self.cpt_status.write().unwrap();
+
+        let mut out = vec![];
+
+        for table in this_level.tables.iter() {
+            if table.size() >= 2 * compact_def.targets.file_size[0] {
+                // file already big, don't include it
+                continue;
```

Comment:
If there are 5 sst files in L0 s1(1MB) s2(1MB) s3(64MB) s4(1MB) s5(1MB), is it will merge s1,s2,s4,s5? If so, it will break the constraint that level0's file is organized by time.

Reply:
We can check line 807 to see how to replace tables. And I think the problem you mentioned will not happen.